components/cdc/src/service.rs

```diff
@@ -473,17 +491,99 @@ impl Service {
             }
         });
 
+        // Create atomic variable to track last flush time
+        let current_time = match std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH) {
+            Ok(duration) => duration.as_secs(),
+            Err(_) => 0,
+        };
+        let last_flush_time = Arc::new(AtomicU64::new(current_time));
+        let last_flush_time_for_forward = last_flush_time.clone();
+        let last_flush_time_for_watchdog = last_flush_time.clone();
+        let peer_for_watchdog = ctx.peer().to_string();
+        
         let peer = ctx.peer();
+        let (cancel_tx, mut cancel_rx) = tokio::sync::oneshot::channel::<()>();
+        let cancel_tx_for_watchdog = cancel_tx;
+        
         ctx.spawn(async move {
             #[cfg(feature = "failpoints")]
             sleep_before_drain_change_event().await;
-            if let Err(e) = event_drain.forward(&mut sink).await {
-                warn!("cdc send failed"; "error" => ?e, "downstream" => peer, "conn_id" => ?conn_id);
-            } else {
-                info!("cdc send closed"; "downstream" => peer, "conn_id" => ?conn_id);
+            tokio::select! {
+                _ = &mut cancel_rx => {
+                    warn!("cdc send cancelled"; "downstream" => peer, "conn_id" => ?conn_id);
+                    let status = RpcStatus::with_message(RpcStatusCode::UNKNOWN, "connection cancelled".to_string());
+                    let _ = sink.fail(status).await;
+                    return;
+                }
+                result = event_drain.forward(&mut sink, Some(&last_flush_time_for_forward)) => {
+                    if let Err(e) = result {
+                        warn!("cdc send failed"; "error" => ?e, "downstream" => peer, "conn_id" => ?conn_id);
+                    } else {
+                        info!("cdc send closed"; "downstream" => peer, "conn_id" => ?conn_id);
+                    }
+                    last_flush_time_for_forward.store(0, Ordering::Relaxed);
+                }
+            }
+        });
+
+        // Add watchdog loop to monitor connection activity using tokio runtime
+        let pool = self.pool.clone();
+        pool.spawn(async move {
+            use std::time::{SystemTime, UNIX_EPOCH, Duration};
+            use tokio::time::interval;
```

Comment:
It's better to place use statements at the top of the function or module rather than inside a block.